---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# curatedMetagenomicData

<!-- badges: start -->
<!-- badges: end -->

curatedMetagenomicData provides standardized, curated human microbiome data for novel analyses. The package includes gene families, marker abundance, marker presence, pathway abundance, pathway coverage, and relative abundance for samples collected from different body sites. The bacterial, fungal, archaeal, and taxonomic abundance of each sample's raw data were calculated with MetaPhlAn3 and the metabolic functional potential was calculated with HUMAnN3. The curated metadata and profiled raw data are available as (Tree)SummarizedExperiment objects.

## Installation

You can install curatedMetagenomicData from Bioconductor with:

```{r, eval = FALSE}
BiocManager::install("curatedMetagenomicData")
```

You can install curatedMetagenomicData from GitHub with:

```{r, eval = FALSE}
BiocManager::install("waldronlab/curatedMetagenomicData", dependencies = TRUE, build_vignettes = TRUE)
```

You can install the last version of curatedMetagenomicData 1 ([v1.20.0](https://github.com/waldronlab/curatedMetagenomicData/releases/tag/v1.20.0)) from GitHub with:

```{r, eval = FALSE}
BiocManager::install("waldronlab/curatedMetagenomicData", dependencies = TRUE, build_vignettes = TRUE, ref = "v1.20.0")
```

## Examples

See the package vignette for more detailed examples.

```{r}
library(curatedMetagenomicData)
```

```{r}
curatedMetagenomicData("AsnicarF_2017")
```

```{r}
curatedMetagenomicData("AsnicarF_2017.relative_abundance")
```

```{r, warning = FALSE}
curatedMetagenomicData("AsnicarF_2017.relative_abundance", dryrun = FALSE)
```

## Contributing

All are welcome to contribute to the curatedMetagenomicData project provided contributions are appropriate. Please see the contributing guide.

## Code of Conduct
  
Please note that the curatedMetagenomicData project is released with a code of conduct. By contributing to this project, you agree to abide by its terms.

---

Pasolli E, Schiffer L, Manghi P, Renson A, Obenchain V, Truong D, Beghini F, Malik F, Ramos M, Dowd J, Huttenhower C, Morgan M, Segata N, Waldron L (2017). Accessible, curated metagenomic data through ExperimentHub. *Nat. Methods*, **14**(11), 1023-1024. ISSN 1548-7091, 1548-7105, doi: [10.1038/nmeth.4468](https://doi.org/10.1038/nmeth.4468).
