---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# curatedMetagenomicData

<!-- badges: start -->
<!-- badges: end -->

The curatedMetagenomicData package provides standardized, curated human microbiome data for novel analyses. It includes gene families, marker abundance, marker presence, pathway abundance, pathway coverage, and relative abundance for samples collected from different body sites. The bacterial, fungal, and archaeal taxonomic abundances for each sample were calculated with MetaPhlAn3 and metabolic functional potential was calculated with HUMAnN3. The manually curated sample metadata and standardized metagenomic data are available as (Tree)SummarizedExperiment objects.

## Versions

To avoid some confusion preemptively - there was *never* a version 2.0.0 of curatedMetagenomicData. Version 1.0.0 was released in 2017 and version 3.0.0 was released in 2021; we skipped directly to version 3.0.0 to align with MetaPhlAn3 and HUMAnN3. Version 3.0.0 features significant changes and, to be sure, they will take some adjustment. While users are encouraged to upgrade to the latest (and greatest) version, the previous version ([1.20.0](https://github.com/waldronlab/curatedMetagenomicData/releases/tag/v1.20.0)) of curatedMetagenomicData will remain installable for the foreseeable future (see below). However, users should consider switching because version 3.0.0 now uses the appropriate modern Bioconductor data structures (`SummarizedExperiment` and `TreeSummarizedExperiment`). As such, all of the previous version 1.0.0 methods that worked on `ExpressionSet` objects are irrelevant and have moved directly to defunct in version 3.0.0. Finally, to avoid additional confusion, the resources pertaining to each release are only available in their respective releases. It would be possible to access them with methods from the `ExperimentHub` package, but doing so is strongly discouraged.

## Installation

You can install the current version of curatedMetagenomicData from Bioconductor with:

```{r, eval = FALSE}
BiocManager::install("curatedMetagenomicData")
```

You can install the current version of curatedMetagenomicData from GitHub with:

```{r, eval = FALSE}
BiocManager::install("waldronlab/curatedMetagenomicData", dependencies = TRUE, build_vignettes = TRUE)
```

You can install the previous version of curatedMetagenomicData from GitHub with:

```{r, eval = FALSE}
BiocManager::install("waldronlab/curatedMetagenomicData", dependencies = TRUE, build_vignettes = TRUE, ref = "v1.20.0")
```

## Examples

Find and access curated metagenomic data using `curatedMetagenomicData()`:

```{r, message = FALSE}
library(curatedMetagenomicData)
```

```{r}
curatedMetagenomicData("AsnicarF_20.+")
```

```{r}
curatedMetagenomicData("AsnicarF_2017.relative_abundance", dryrun = FALSE)
```

```{r}
curatedMetagenomicData("AsnicarF_20.+.relative_abundance", dryrun = FALSE, counts = TRUE)
```

See the package vignette for more detailed examples.

## Contributing

All are welcome to contribute to the curatedMetagenomicData package provided contributions are appropriate. Please see the contributing guide.

## Code of Conduct
  
Please note that the curatedMetagenomicData package is released with a code of conduct. By contributing, you agree to abide by its terms.

---

Pasolli E, Schiffer L, Manghi P, Renson A, Obenchain V, Truong D, Beghini F, Malik F, Ramos M, Dowd J, Huttenhower C, Morgan M, Segata N, Waldron L (2017). Accessible, curated metagenomic data through ExperimentHub. *Nat. Methods*, **14**(11), 1023-1024. ISSN 1548-7091, 1548-7105, doi: [10.1038/nmeth.4468](https://doi.org/10.1038/nmeth.4468).
