---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# curatedMetagenomicData

<!-- badges: start -->
<!-- badges: end -->

The curatedMetagenomicData package provides microbial taxonomic, functional, and gene marker abundance for samples collected from different body sites.

## Installation

You can install curatedMetagenomicData from Bioconductor with:

```{r, eval = FALSE}
BiocManager::install("curatedMetagenomicData")
```

You can install curatedMetagenomicData from GitHub with:

```{r, eval = FALSE}
BiocManager::install("waldronlab/curatedMetagenomicData", dependencies = TRUE, build_vignettes = TRUE)
```

You can install the last version of curatedMetagenomicData 1 ([v1.20.0](https://github.com/waldronlab/curatedMetagenomicData/releases/tag/v1.20.0)) from GitHub with:

```{r, eval = FALSE}
BiocManager::install("waldronlab/curatedMetagenomicData", dependencies = TRUE, build_vignettes = TRUE, ref = "v1.20.0")
```

## Command-line installation

You must first install the R package, see instructions above. Then you can place the command-line tool on your path as follows:

```
CMDLIB=$(R -e 'find.package("curatedMetagenomicData")' | grep -o -e "/.*\w")/exec
echo -e "export PATH=\$PATH:$CMDLIB" >> ~/.bashrc
```

### Command-line usage

```
Usage:
  curatedMetagenomicData [--metadata] [--counts] [--dryrun] [<NAME>]...
  curatedMetagenomicData [-mcd] [<NAME>]...
  curatedMetagenomicData -l | --list
  curatedMetagenomicData -h | --help
```

## Examples

See the package vignette for more detailed examples.

```{r, message=FALSE}
library(curatedMetagenomicData)
```

```{r, message=FALSE}
curatedMetagenomicData("AsnicarF_2017")
```

```{r, message=FALSE}
curatedMetagenomicData("AsnicarF_2017.relative_abundance")
```

```{r, message=FALSE}
curatedMetagenomicData("AsnicarF_2017.relative_abundance", dryrun = FALSE)
```

## Contributing

All are welcome to contribute to the curatedMetagenomicData project provided contributions are appropriate. Please see the contributing guide.

## Code of Conduct
  
Please note that the curatedMetagenomicData project is released with a code of conduct. By contributing to this project, you agree to abide by its terms.

---

Pasolli E, Schiffer L, Manghi P, Renson A, Obenchain V, Truong D, Beghini F, Malik F, Ramos M, Dowd J, Huttenhower C, Morgan M, Segata N, Waldron L (2017). Accessible, curated metagenomic data through ExperimentHub. *Nat. Methods*, **14**(11), 1023-1024. ISSN 1548-7091, 1548-7105, doi: [10.1038/nmeth.4468](https://doi.org/10.1038/nmeth.4468).
