---
title: "curatedMetagenomicData"
author:
- name: Lucas Schiffer, MPH
  affiliation: 
  - Section of Computational Biomedicine, Boston University School of Medicine,
    Boston, MA, U.S.A.
  email: schifferl@bu.edu
package: curatedMetagenomicData
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{curatedMetagenomicData}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r, include = FALSE}
library(curatedMetagenomicData)
```


# Taxonomy-aware analysis using phyloseq

Relative abundance data is delivered as a TreeSummarizedExperiment-class object, which allows
operations using functions from the [TreeSummarizedExperiment](https://www.bioconductor.org/packages/release/bioc/html/TreeSummarizedExperiment.html)
package. This data can be converted to a phyloseq-class object to work with functions from the
[phyloseq](https://bioconductor.org/packages/release/bioc/html/phyloseq.html) package.
Conversion from TreeSummarizedExperiment to phyloseq is supported by other packages, such as the [mia](https://bioconductor.org/packages/devel/bioc/html/mia.html)
package in Bioconductor.

```{r}
suppressMessages({
  library(mia)
  library(phyloseq)
})

```

An example using the relative abundance data from the LomanNJ_2013 study:

```{r, message=FALSE, warning=FALSE}
loman.tse <- curatedMetagenomicData(".+LomanNJ_2013.relative.abundance", dryrun = FALSE, counts = FALSE)[[1]]
loman.tse <- loman.tse[rowData(loman.tse)$kingdom == "Bacteria",]
class(loman.tse)
```

Convert to phyloseq-class object:

```{r}
loman.pseq <- mia::makePhyloseqFromTreeSummarizedExperiment(loman.tse, abund_values = "relative_abundance")
loman.pseq
class(loman.pseq)
```

## Phylogenetic trees and UniFrac distances

Since the phylogenetic tree from MetaPhlAn3 is already integrated in the data, computing UniFrac distances can be achieved with:

```{r}
wt <- UniFrac(loman.pseq, weighted = TRUE, normalized = TRUE, parallel = FALSE, fast = TRUE)
plot(hclust(wt), main = "LomanNJ_2013 - Weighted UniFrac distances")
```

## Components of a phyloseq object

This phyloseq objects contain 4 components with extractor functions hinted at by its show method:

```{r}
loman.pseq
```

`otu_table()` returns the Operational Taxanomic Unit (OTU) table. 

```{r}
otu_table(loman.pseq)[1:5, 1:5]
```

`sample_data()` returns the sample metadata:

```{r}
sample_data(loman.pseq)[1:5, 1:5]
```

`tax_table()`returns the taxonomy information:

```{r}
tax_table(loman.pseq)[1:5, 1:5]
```

`phy_tree()` returns the phylogenetic tree:

```{r}
phy_tree(loman.pseq)
```

## Subsetting / pruning

Subset can be applied on the phyloseq object to obtain information of specific taxa levels:


```{r}
rank_names(loman.pseq)

```

To keep only phylum-level data:

```{r}
loman.bd = subset_taxa(loman.pseq, phylum == "Bacteroidetes")
loman.bd
head(taxa_names( loman.bd ))
```

## Advanced pruning

The phyloseq package provides advanced pruning of taxa. The follwing code keeps only
taxa that are among the most abundant 5% in at least ten samples:

```{r}
keepotu = genefilter_sample(loman.pseq, filterfun_sample(topp(0.05)), A = 10)
summary(keepotu)
subset_taxa(loman.pseq, keepotu)

```

Note that phyloseq also provides `topk()` for selecting the most abundant k taxa,
and other functions for advanced pruning of taxa.

## Taxonomy heatmap

The phyloseq package provides the `plot_heatmap()` function to create heatmaps using a variety of built-in
dissimilarity metrics for clustering. Here, we apply the same abundance filter as above. This function supports 
a large number of distance and ordination methods, here we use Bray-Curtis dissimilarity for distance and PCoA
as the ordination method for organizing the heatmap.

```{r, message=FALSE, warning=FALSE}
loman.filt = subset_taxa(loman.pseq, keepotu)
plot_heatmap(loman.filt, method="PCoA", distance="bray") +
  ggplot2::theme(axis.text.y = ggplot2::element_blank(),
                 axis.ticks.y = ggplot2::element_blank())
```


## Taxonomy histogram

Here we plot the top 20 most abundant species (not strains), defined by the sum of abundance across all samples in the dataset:

```{r, fig.height=8.5}
par(mar = c(20, 4, 0, 0) + 0.15) #increase margin size on the bottom
barplot(sort(taxa_sums(loman.pseq), TRUE)[1:20] / nsamples(loman.pseq),
        ylab = "Total counts", las = 2)


```

## Alpha diversity estimation

The phyloseq package calculates numerous alpha diversity measures. Here we compare three diversity
in the species-level data,stratifying by stool texture:

```{r, message=FALSE, warning=FALSE}
alphas = c("Shannon", "Simpson", "InvSimpson")
plot_richness(loman.pseq, "stool_texture", measures = alphas)
```

Let’s compare these three alpha diversity measures:

```{r, message=FALSE, warning=FALSE}
pairs(estimate_richness(loman.pseq, measures = alphas))

```

## Beta diversity / dissimilarity clustering

Numerous beta diversity/dissimilarity distances are provided by the `distance()` function when provided a phyloseq object,
and these can be used for any kind of clustering or classification scheme. For example, here is a hierarchical
clustering dendrogram produced by the `hclust()` from the base R stats package with “Ward” linkage:


```{r}
mydist = distance(loman.pseq, method="bray")
myhclust = hclust( mydist )
plot(myhclust, main="LomanNJ_2013 - Bray-Curtis Dissimilarity", 
     method="ward.D", xlab="Samples", sub = "")
```

## Ordination analysis

The phyloseq package provides a variety of ordination methods, with convenient options
for labeling points. Here is a Principal Coordinates Analysis plot of species-level
taxa from the Loman dataset, using Bray-Curtis distance:

```{r}
ordinated_taxa = ordinate(loman.pseq, method="PCoA", distance="bray")
plot_ordination(loman.pseq, ordinated_taxa, color="stool_texture", 
                title = "LomanNJ_2013 - Bray-Curtis Principal Coordinates Analysis")
```

```{r}
plot_scree(ordinated_taxa, title="LomanNJ_2013 - Screeplot")

```
